### PR TITLE
Added utility to count memory reads/written in Inductor

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89171
* __->__ #89203
* #89128

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -353,6 +353,32 @@ def codegen(self):
...
+                    total_numel += numel
+                elif name in V.graph.graph_inputs:
+                    buf = V.graph.graph_inputs[name]
+                    numel = V.graph.sizevars.size_hint(sympy_product(buf.get_size()))
+                    print(name, numel)
+                    total_numel += numel
+            return total_numel
+        total_numel = 0
+        for node in self.scheduler.nodes:
+            numel = get_read_write_buffers_sizes(node)
```
Comment (ngimel) [inline torch/_inductor/graph.py:378]: that's a nice place to put it in :-)

Comment (Chillee) [reply torch/_inductor/graph.py:378]: I'm probably just going to make it a separate method on Graph.

File: test/inductor/test_perf.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    return torch.cat([x, x.cos()])
+
+
+def count_numel(f, *args):
+    """
+    Assumes all inputs are fp32
+    """
+    metrics.reset()
+    torch._dynamo.optimize("count_bytes_inductor")(f)(*args)
+    return str(metrics.num_bytes_accessed // 1)
```
Comment (ngimel) [inline test/inductor/test_perf.py:28]: nit: it's not bytes, it's numel

**Review by ngimel — COMMENTED** [review]
Nice!
Why did you decide to make it a separate backend?

**Comment by Chillee** [comment]
I made it a separate backend since I wanted to be able to run this test *without* actually compiling any Triton kernels. So that make it difficult to reuse `compile_fx` as is.

**Review by jansel — APPROVED** [review]

File: test/inductor/test_perf.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+
+def count_numel(f, *args):
+    """
+    Assumes all inputs are fp32
+    """
+    metrics.reset()
+    torch._dynamo.optimize("count_bytes_inductor")(f)(*args)
+    print(metrics.nodes_num_elem)
+    return str(metrics.num_bytes_accessed // 4)
```
Comment (ngimel) [inline test/inductor/test_perf.py:36]: works for all datatypes as long as their size is 4.

**Review by ngimel — APPROVED** [review]

File: test/inductor/test_perf.py (hunk 3 above)
Comment (Chillee) [reply test/inductor/test_perf.py:36]: Well, I just want to deal with "number of elements" rather than bytes when writing tests lol.

**Pull request closed without merging**